src/Features/Core/Portable/AddImport/AbstractAddImportFeatureService.cs

```diff
@@ -475,5 +470,111 @@ private static bool NotNull(SymbolReference reference)
+                ? solution.Workspace.Services.GetService<ISymbolSearchService>()
+                : null;
+
+            var packageSources = symbolSearchService != null && searchNuGetPackages
+                ? GetPackageSources(document)
+                : ImmutableArray<PackageSource>.Empty;
+
+            return GetFixesForDiagnosticsAsync(document, span, diagnostics, symbolSearchService, searchNuGetPackages, packageSources, cancellationToken);
+        }
+
+        public async Task<ImmutableArray<(Diagnostic Diagnostic, ImmutableArray<AddImportFixData> Fixes)>> GetFixesForDiagnosticsAsync(
+            Document document, Text.TextSpan span, ImmutableArray<Diagnostic> diagnostics,
+            ISymbolSearchService symbolSearchService, bool searchReferenceAssemblies, 
+            ImmutableArray<PackageSource> packageSources, CancellationToken cancellationToken)
+        {
+            var addImportService = document.GetLanguageService<IAddImportFeatureService>();
+
+            // We might have multiple different diagnostics covering the same span.  Have to
+            // process them all as we might produce different fixes for each diagnostic.
+
+            var documentOptions = await document.GetOptionsAsync(cancellationToken).ConfigureAwait(false);
+            var placeSystemNamespaceFirst = documentOptions.GetOption(GenerationOptions.PlaceSystemNamespaceFirst);
+
+            var fixesForDiagnosticBuilder = ArrayBuilder<(Diagnostic, ImmutableArray<AddImportFixData>)>.GetInstance();
+
+            foreach (var diagnostic in diagnostics)
+            {
+                var fixes = await addImportService.GetFixesAsync(
+                    document, span, diagnostic.Id, placeSystemNamespaceFirst,
+                    symbolSearchService, searchReferenceAssemblies,
+                    packageSources, cancellationToken).ConfigureAwait(false);
+
+                fixesForDiagnosticBuilder.Add((diagnostic, fixes));
+            }
+
+            return fixesForDiagnosticBuilder.ToImmutableAndFree();
+        }
+
+        public ImmutableArray<CodeAction> GetCodeActionsForFixes(Document document, ImmutableArray<AddImportFixData> fixes, IPackageInstallerService installerService = null)
+        {
+            var codeActionsBuilder = ArrayBuilder<CodeAction>.GetInstance();
+
+            foreach (var fix in fixes)
+            {
+                if (TryCreateCodeAction(document, fix, out var codeAction, installerService))
+                {
+                    codeActionsBuilder.AddIfNotNull(codeAction);
+
+                    if (codeActionsBuilder.Count >= MaxResults)
+                    {
+                        break;
+                    }
+                }
+            }
+
+            return codeActionsBuilder.ToImmutableAndFree();
+        }
+
+        public bool TryCreateCodeAction(Document document, AddImportFixData fixData, out CodeAction codeAction, IPackageInstallerService installerService = null)
```

Comment:
i do not like this optional param :)

Reply:
why not just hav ethis return the CodeAction.  It can return null for failure (implied by 'Try').

Reply:
You don't like it being optional or you don't like it following an `out` parameter? =) Happy to return `CodeAction` instead.

Reply:
Both... :)  Can you make it non-optional and make it explicit when people are passing it, or are passing null.

The problem with optional parameters is that it is never clear from the callsite if the omission was intentional, or indicates some sort of bug.